File: clients/drive9-js/tests/pathfilter.test.ts
import { describe, expect, it } from "vitest";
import { compile, compileAll, match, matchExcluded, newMatcher, validate, matchPattern } from "../src/pathfilter.js";

describe("pathfilter compile", () => {
  it("compiles subpath form **/x/**", () => {
    const p = compile("**/node_modules/**");
    expect(matchPattern(p, "proj/src/node_modules/react/index.js")).toBe(true);
    expect(matchPattern(p, "node_modules/react/index.js")).toBe(true);
    expect(matchPattern(p, "proj/src/app/main.go")).toBe(false);
  });

  it("compiles prefix form prefix/**", () => {
    const p = compile("dist/**");
    expect(matchPattern(p, "dist/index.js")).toBe(true);
    expect(matchPattern(p, "dist/a/b/c.js")).toBe(true);
    expect(matchPattern(p, "src/index.js")).toBe(false);
  });

  it("compiles exact/glob form", () => {
    const p = compile("*.log");
    expect(matchPattern(p, "app.log")).toBe(true);
    expect(matchPattern(p, "app.txt")).toBe(false);
    const exact = compile("build");
    expect(matchPattern(exact, "build")).toBe(true);
    expect(matchPattern(exact, "build/dist")).toBe(false);
  });
});

describe("pathfilter recursive glob segments", () => {
  it.each([
    ["**/*-wal", "repro.db-wal", true],
    ["**/*-wal", "/issue-validation/run/case/repro.db-wal", true],
    ["**/repro.db-wal", "/issue-validation/run/case/repro.db-wal", true],
    ["**/*-wal", "proj/other.sqlite-wal", true],
    ["**/*-wal", "proj/repro.db", false],
    ["**/*-wal", "proj/repro.db-shm", false],
    ["**/*-wal", "proj/repro.db-wal.bak", false],
    ["**/*-wal", "/snapshots-wal/data.bin", false],
    ["**/*-wal", "/proj/snapshots-wal/sub/data.bin", false],
    ["**/*-wal", "/snapshots-wal/repro.db-wal", true],
    ["**/cache-*/**", "proj/cache-build/objects/a", true],
    ["**/cache-*", "proj/cache-build/objects/a", false],
    ["**/cache-*", "proj/cache-build", true],
    ["**/cache-*/*.log", "proj/cache-build/output.log", true],
    ["**/cache-*/*.log", "proj/cache-build/nested/output.log", false],
    ["**/cache-*/*.log", "proj/cache-build/output.log/data.bin", false],
    ["**/cache-*/*.log", "cache-first/miss/cache-second/output.log", true],
    ["**/db?.wal", "proj/db1.wal", true],
    ["**/db?.wal", "proj/db12.wal", false],
    ["**/*.[Tt]xt", "proj/a.Txt", true],
    ["**/*.[^Tt]xt", "proj/a.bxt", true],
    ["**/*.[^Tt]xt", "proj/a.txt", false],
    ["**/[cache]/**", "proj/[cache]/item", true],
    ["**/[z-a]/**", "proj/[z-a]/item", true],
    ["**/[z-a]/**", "proj/other/item", false],
    ["**/db?.wal", "proj/db\u{1f600}.wal", true],
    ["**/db??.wal", "proj/db\u{1f600}.wal", false],
    ["**/db[\u{1f600}].wal", "proj/db\u{1f600}.wal", true],
    ["**/db[\u{1f600}]?.wal", "proj/db\u{1f600}.wal", false],
    ["**/db[\u{1f600}-\u{1f64f}].wal", "proj/db\u{1f609}.wal", true],
    ["**/db[^\u{1f600}].wal", "proj/db\u{1f609}.wal", true],
    ["**/db[^a].wal", "proj/db\u{1f600}.wal", true],
    ["**/db[^a]?.wal", "proj/db\u{1f600}.wal", false],
    ["**/db?.wal", "proj/db\n.wal", true],
    ["**/db*.wal", "proj/db\n\u{1f600}.wal", true],
    ["**/cafe\u0301*.txt", "proj/caf\u00e91.txt", true],
    ["**/caf\u00e9*.txt", "proj/cafe\u03011.txt", true],
    ["**/caf?.txt", "proj/cafe\u0301.txt", true],
    ["**/caf[e\u0301].txt", "proj/caf\u00e9.txt", true],
    ["**/*?", "proj/\u{1f600}", true],
    ["**/*??", "proj/\u{1f600}", false],
    ["**/*??", "proj/\u{1f600}a", true],
    ["**/*[^a][^a]", "proj/\u{1f600}", false],
    ["**/*[\ufffd]", "proj/\u{1f600}", false],
    ["**/*??x", "proj/\u{1f600}x", false],
    ["**/*a*??", "proj/a\u{1f600}", false],
    ["**/*a*??", "proj/a\u{1f600}a", true],
    ["**/[*]*??", "proj/*\u{1f600}", false],
    ["**/*-wal", "proj/\u{1f600}-wal", true],
    ["*??", "\u{1f600}", false],
    ["*??/file", "\u{1f600}/file", false],
    ["*?/file", "\u{1f600}/file", true],
    ["*?file", "\u{1f600}/file", false],
    ["**/[a-]/**", "proj/a/item", false],
    ["**/[a-]/**", "proj/[a-]/item", true],
    ["**/[a-]/**", "proj/-/item", false],
    ["**/[-a]/**", "proj/a/item", false],
    ["**/[]/**", "proj/a/item", false],
    ["**/[^]/**", "proj/a/item", false],
    ["**/[]a]/**", "proj/a/item", false],
    ["**/[abc/**", "proj/[/item", false],
    ["**/[a-b-c]/**", "proj/-/item", false],
    ["**/[z-aa]/**", "proj/a/item", true],
    ["**/[^z-a]/**", "proj/a/item", true],
    ["**/[[]/**", "proj/[/item", true],
    ["**/db?].wal", "proj/db1].wal", true],
    ["**/[abc", "proj/[abc/secret", true],
    ["**/[a-]", "proj/[a-]/secret", true],
    ["**/[abc/name", "proj/[abc/name/secret", true],
    ["**/[a/b]", "proj/[a/b]/secret", true],
    ["**/prefix*/[abc", "proj/prefix1/[abc", true],
    ["**/prefix*/[abc", "proj/prefix1/[abc/secret", false],
    ["**/[z-a]", "proj/[z-a]/secret", false],
  ])("%s matching %s returns %s", (pattern, path, want) => {
    expect(matchPattern(compile(pattern), path)).toBe(want);
  });
});

describe("pathfilter matcher", () => {
  it("include whitelist + exclude", () => {
    const m = newMatcher({ include: ["proj/**", "go.mod"], exclude: ["**/vendor/**"] });
    expect(match(m, "proj/src/main.go")).toBe(true);
    expect(match(m, "go.mod")).toBe(true);
    expect(match(m, "proj/vendor/foo.go")).toBe(false);
    expect(match(m, "README.md")).toBe(false);
  });

  it("no include accepts all non-excluded", () => {
    const m = newMatcher({ exclude: ["**/.git/**", "**/node_modules/**"] });
    expect(match(m, "proj/src/app.go")).toBe(true);
    expect(match(m, "proj/node_modules/react/x.js")).toBe(false);
    expect(match(m, "proj/.git/HEAD")).toBe(false);
  });

  it("override restores excluded path", () => {
    const m = newMatcher({
      exclude: ["**/node_modules/**"],
      override: ["proj/node_modules/.package-lock.json"],
    });
    expect(match(m, "proj/node_modules/react/x.js")).toBe(false);
    expect(match(m, "proj/node_modules/.package-lock.json")).toBe(true);
    expect(match(m, "proj/src/app.go")).toBe(true);
  });

  it("exclude wins over include without override", () => {
    const m = newMatcher({ include: ["proj/**"], exclude: ["**/vendor/**"] });
    expect(match(m, "proj/vendor/foo.go")).toBe(false);
    expect(match(m, "proj/src/foo.go")).toBe(true);
  });

  it("empty matcher matches all", () => {
    const m = newMatcher({});
    expect(match(m, "anything/here.go")).toBe(true);
  });
});

describe("pathfilter validate", () => {
  it.each(["**/[\\-]/**", "**/[\\]]/**"])("rejects backslashes in %s", (pattern) => {
    expect(() => compile(pattern)).toThrow("backslash");
    expect(validate([pattern])).toBeInstanceOf(Error);
    expect(compileAll([pattern])).toEqual([]);
  });

  it("returns null for valid patterns", () => {
    expect(validate(["dist/**", "*.log"], ["**/x/**"])).toBeNull();
  });
});

describe("pathfilter compileAll", () => {
  it("skips blank entries", () => {
    expect(compileAll(["", "  ", "dist/**"]).length).toBe(1);
  });
});

describe("pathfilter matchExcluded", () => {
  it("does not prune directories selected only by recursive file globs", () => {
    const m = newMatcher({ exclude: ["**/*-wal", "**/cache-*/**", "**/vendor", "**/[abc"] });
    expect(match(m, "snapshots-wal")).toBe(false);
    expect(matchExcluded(m, "snapshots-wal")).toBe(false);
    expect(match(m, "snapshots-wal/data.bin")).toBe(true);
    expect(match(m, "snapshots-wal/repro.db-wal")).toBe(false);
    expect(matchExcluded(m, "cache-build")).toBe(true);
    expect(matchExcluded(m, "proj/vendor")).toBe(true);
    expect(matchExcluded(m, "proj/[abc")).toBe(true);
    expect(match(m, "proj/[abc/secret")).toBe(false);
  });

  it("returns true only for exclude-matched paths not restored by override", () => {
    const m = newMatcher({
      include: ["src/app.go"],
      exclude: ["**/node_modules/**", "**/.git/**"],
      override: ["proj/node_modules/.package-lock.json"],
    });
    expect(matchExcluded(m, "proj/node_modules/react")).toBe(true);
    expect(matchExcluded(m, "proj/.git")).toBe(true);
    // Override restores — not excluded.
    expect(matchExcluded(m, "proj/node_modules/.package-lock.json")).toBe(false);
    // A directory that only fails the include whitelist is NOT excluded —
    // pruning it would drop its child src/app.go (the B2 bug).
    expect(matchExcluded(m, "src")).toBe(false);
    expect(matchExcluded(m, "src/util.go")).toBe(false);
  });
});

describe("pathfilter glob ? and [abc]", () => {
  it("supports ? as single non-separator char", () => {
    const p = compile("a?.go");
    expect(matchPattern(p, "ab.go")).toBe(true);
    expect(matchPattern(p, "abc.go")).toBe(false); // ? is single char
    expect(matchPattern(p, "a/.go")).toBe(false); // ? does not cross separators
  });

  it("supports character classes", () => {
    const p = compile("*.[Tt]xt");
    expect(matchPattern(p, "a.txt")).toBe(true);
    expect(matchPattern(p, "a.Txt")).toBe(true);
    expect(matchPattern(p, "a.go")).toBe(false);
  });

  it("supports negated character classes", () => {
    const p = compile("*.[^Tt]xt");
    expect(matchPattern(p, "a.txt")).toBe(false);
    expect(matchPattern(p, "a.bxt")).toBe(true);
  });
});

File: pkg/pathfilter/pathfilter_test.go
package pathfilter

import (
	"testing"
)

func TestCompileThreeForms(t *testing.T) {
	cases := []struct {
		name    string
		pattern string
		path    string
		want    bool
	}{
		// **/x/** subpath form
		{"subpath-match", "**/node_modules/**", "proj/src/node_modules/react/index.js", true},
		{"subpath-match-root-level", "**/node_modules/**", "node_modules/react/index.js", true},
		{"subpath-no-match", "**/node_modules/**", "proj/src/app/main.go", false},
		{"subpath-nested", "**/.git/**", "proj/.git/refs/heads/main", true},
		{"subpath-bare-dir", "**/dist", "proj/dist", true},
		{"subpath-bare-dir-no-trailing", "**/dist", "proj/dist/index.js", true},
		// prefix/** form
		{"prefix-match", "dist/**", "dist/index.js", true},
		{"prefix-nested", "dist/**", "dist/a/b/c.js", true},
		{"prefix-no-match", "dist/**", "src/index.js", false},
		{"prefix-exact-dir", "dist/**", "dist", true},
		// exact/glob form
		{"exact-match", "build", "build", true},
		{"exact-no-match", "build", "build/dist", false},
		{"glob-match", "*.log", "app.log", true},
		{"glob-no-match", "*.log", "app.txt", false},
		{"glob-dir-level", "*.log", "logs/app.log", false}, // path.Match against full path
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			p, err := Compile(tc.pattern)
			if err != nil {
				t.Fatalf("Compile(%q) error: %v", tc.pattern, err)
			}
			if got := p.Match(tc.path); got != tc.want {
				t.Fatalf("Pattern(%q).Match(%q) = %v, want %v", tc.pattern, tc.path, got, tc.want)
			}
		})
	}
}

func TestValidateRejectsInvalid(t *testing.T) {
	// pathutil.Canonicalize rejects backslashes and .. segments.
	if err := Validate([]string{"**/../etc/**"}); err == nil {
		t.Fatal("Validate accepted a pattern containing ..")
	}
}

func TestMatcherIncludeExclude(t *testing.T) {
	// Include whitelist: everything under proj/** plus go.mod at root.
	// Exclude: vendor subtree and *_test.go exact files (glob, root-level only).
	m := NewMatcher(
		[]string{"proj/**", "go.mod"},
		[]string{"**/vendor/**", "*_test.go"},
		nil,
	)
	cases := []struct {
		path string
		want bool
	}{
		{"proj/src/main.go", true},            // matches include, not exclude
		{"proj/src/server/server.go", true},   // matches include, not exclude
		{"go.mod", true},                       // exact include match
		{"proj/vendor/foo.go", false},          // matches include but excluded (vendor)
		{"proj/main_test.go", true},           // *_test.go exclude is root-level glob only
		{"main_test.go", false},                // excluded by root-level *_test.go glob
		{"README.md", false},                   // not in include whitelist
	}
	for _, tc := range cases {
		t.Run(tc.path, func(t *testing.T) {
			if got := m.Match(tc.path); got != tc.want {
				t.Fatalf("Match(%q) = %v, want %v", tc.path, got, tc.want)
			}
		})
	}
}

func TestMatcherNoIncludeAcceptsAll(t *testing.T) {
	m := NewMatcher(nil, []string{"**/.git/**", "**/node_modules/**"}, nil)
	if !m.Match("proj/src/app.go") {
		t.Fatal("non-excluded path should match when include is empty")
	}
	if m.Match("proj/node_modules/react/index.js") {
		t.Fatal("excluded path should not match")
	}
	if m.Match("proj/.git/HEAD") {
		t.Fatal(".git should be excluded")
	}
}

func TestMatcherOverrideRestoresExcluded(t *testing.T) {
	// Simulate profile coding-agent: exclude node_modules, but override
	// restores a specific project's node_modules manifest.
	m := NewMatcher(
		nil,
		[]string{"**/node_modules/**"},
		[]string{"proj/node_modules/.package-lock.json"},
	)
	if m.Match("proj/node_modules/react/index.js") {
		t.Fatal("excluded node_modules should be dropped")
	}
	if !m.Match("proj/node_modules/.package-lock.json") {
		t.Fatal("override should restore the package-lock")
	}
	if !m.Match("proj/src/app.go") {
		t.Fatal("non-excluded path should pass")
	}
}

func TestMatcherExcludeOverridesInclude(t *testing.T) {
	// If a path matches both include and exclude, exclude wins (unless override).
	m := NewMatcher([]string{"proj/**"}, []string{"**/vendor/**"}, nil)
	if m.Match("proj/vendor/foo.go") {
		t.Fatal("exclude must win over include when no override present")
	}
	if !m.Match("proj/src/foo.go") {
		t.Fatal("included non-excluded path should match")
	}
}

func TestCodingAgentDefaultPatterns(t *testing.T) {
	patterns := []string{
		"**/.git/**", "**/.hg/**", "**/.svn/**", "**/node_modules/**",
		"**/.pnpm-store/**", "**/target/**", "**/dist/**", "**/build/**",
		"**/coverage/**", "**/tmp/**", "**/.tmp/**", "**/.tmp-api-extractor/**",
		"**/.cache/**", "**/.turbo/**", "**/.next/cache/**", "**/.vitepress/cache/**",
		"**/.gradle/**", "**/.venv/**", "**/__pycache__/**", "**/.pytest_cache/**",
		"**/.mypy_cache/**", "**/.ruff_cache/**",
	}
	m := NewMatcher(nil, patterns, nil)
	drop := []string{
		".git/HEAD", "proj/.git/config", "node_modules/react/index.js",
		"proj/dist/bundle.js", "proj/build/output.o", "proj/.venv/bin/python",
		"proj/__pycache__/foo.cpython-311.pyc", "proj/.next/cache/abc",
		"proj/.pytest_cache/v/cache/lastfailed",
	}
	for _, p := range drop {
		if m.Match(p) {
			t.Fatalf("coding-agent pattern should drop %q", p)
		}
	}
	keep := []string{"proj/src/main.go", "README.md", "proj/go.mod", "proj/.gitignore"}
	for _, p := range keep {
		if !m.Match(p) {
			t.Fatalf("coding-agent pattern should keep %q", p)
		}
	}
}

func TestEmptyMatcherMatchesAll(t *testing.T) {
	m := NewMatcher(nil, nil, nil)
	if !m.Match("anything/here.go") {
		t.Fatal("empty matcher should match all")
	}
	if !m.Match("") {
		t.Fatal("empty matcher should match empty path")
	}
}

func TestCompileAllSkipsBlankAndInvalid(t *testing.T) {
	got := CompileAll([]string{"", "  ", "dist/**", "**/../bad/**"})
	if len(got) != 1 {
		t.Fatalf("CompileAll kept %d patterns, want 1 (only dist/**)", len(got))
	}
	if got[0].Raw() != "dist/**" {
		t.Fatalf("CompileAll kept wrong pattern: %q", got[0].Raw())
	}
}

func TestMatchExcluded(t *testing.T) {
	m := NewMatcher(
		[]string{"src/app.go"}, // include whitelist (only this leaf)
		[]string{"**/node_modules/**", "**/.git/**"},
		[]string{"proj/node_modules/.package-lock.json"}, // override
	)
	// Excluded subtrees.
	if !m.MatchExcluded("proj/node_modules/react") {
		t.Fatal("node_modules/react should be MatchExcluded (prune subtree)")
	}
	if !m.MatchExcluded("proj/.git") {
		t.Fatal(".git should be MatchExcluded")
	}
	// Override restores — NOT excluded.
	if m.MatchExcluded("proj/node_modules/.package-lock.json") {
		t.Fatal("override should make package-lock.json not MatchExcluded")
	}
	// A directory that only fails the include whitelist is NOT MatchExcluded —
	// pruning it would drop its children (the B2 bug).
	if m.MatchExcluded("src") {
		t.Fatal("src (only failing include) must NOT be MatchExcluded — its child src/app.go is included")
	}
	// Leaf that fails include is not "excluded" in the prune sense either.
	if m.MatchExcluded("src/util.go") {
		t.Fatal("src/util.go is dropped by include, not by exclude — not MatchExcluded")
	}
}

func TestGlobQuestionAndCharClass(t *testing.T) {
	// Character class form: use direct pattern matching (compile + Match)
	// so the matcher's include/exclude semantics don't interfere.
	p, err := Compile("*.[Tt]xt")
	if err != nil {
		t.Fatalf("Compile *.[Tt]xt: %v", err)
	}
	if !p.Match("a.txt") {
		t.Fatal("*.[Tt]xt should match a.txt")
	}
	if !p.Match("a.Txt") {
		t.Fatal("*.[Tt]xt should match a.Txt")
	}
	if p.Match("a.go") {
		t.Fatal("*.[Tt]xt should not match a.go")
	}
	// ? matches a single non-separator char.
	pq, err := Compile("a?.go")
	if err != nil {
		t.Fatalf("Compile a?.go: %v", err)
	}
	if !pq.Match("ab.go") {
		t.Fatal("a?.go should match ab.go")
	}
	if pq.Match("abc.go") {
		t.Fatal("a?.go should not match abc.go (? is single char)")
	}
	if pq.Match("a/.go") {
		t.Fatal("a?.go should not match a/.go (? does not cross separators)")
	}
	// Negated class.
	pn, err := Compile("*.[^Tt]xt")
	if err != nil {
		t.Fatalf("Compile *.[^Tt]xt: %v", err)
	}
	if pn.Match("a.txt") {
		t.Fatal("*.[^Tt]xt should not match a.txt")
	}
	if !pn.Match("a.bxt") {
		t.Fatal("*.[^Tt]xt should match a.bxt")
	}
}